source/credentials_provider_sts_web_identity.c

```diff
@@ -470,87 +470,129 @@ static void s_on_stream_complete_fn(struct aws_http_stream *stream, int error_co
 
-    struct aws_http_header accept_header = {
-        .name = aws_byte_cursor_from_string(s_accept_header),
-        .value = aws_byte_cursor_from_string(s_accept_header_value),
+    struct aws_http_header content_len_header = {
+        .name = s_content_length,
+        .value = aws_byte_cursor_from_c_str(content_length),
     };
-    if (aws_http_message_add_header(request, accept_header)) {
+
+    if (aws_http_message_add_header(request, content_len_header)) {
         goto on_error;
     }
 
-    struct aws_http_header user_agent_header = {
-        .name = aws_byte_cursor_from_string(s_user_agent_header),
-        .value = aws_byte_cursor_from_string(s_user_agent_header_value),
-    };
-    if (aws_http_message_add_header(request, user_agent_header)) {
+    if (aws_http_message_add_header(request, s_content_type_header)) {
         goto on_error;
     }
 
-    struct aws_http_header keep_alive_header = {
-        .name = aws_byte_cursor_from_string(s_h1_0_keep_alive_header),
-        .value = aws_byte_cursor_from_string(s_h1_0_keep_alive_header_value),
-    };
-    if (aws_http_message_add_header(request, keep_alive_header)) {
+    if (aws_http_message_add_header(request, s_host_header)) {
         goto on_error;
     }
 
-    if (aws_http_message_set_request_path(request, *uri)) {
+    if (aws_http_message_add_header(request, s_api_version_header)) {
         goto on_error;
     }
 
-    if (aws_http_message_set_request_method(request, aws_http_method_get)) {
+    if (aws_http_message_add_header(request, s_accept_header)) {
         goto on_error;
     }
 
-    sts_web_identity_user_data->request = request;
+    if (aws_http_message_add_header(request, s_user_agent_header)) {
+        goto on_error;
+    }
+
+    if (aws_http_message_add_header(request, s_keep_alive_header)) {
+        goto on_error;
+    }
+
+    input_stream = aws_input_stream_new_from_cursor(user_data->allocator, body_cursor);
+    if (!input_stream) {
+        goto on_error;
+    }
+
+    aws_http_message_set_body_stream(request, input_stream);
+
+    if (aws_http_message_set_request_path(request, s_path)) {
+        goto on_error;
+    }
+
+    if (aws_http_message_set_request_method(request, aws_http_method_post)) {
+        goto on_error;
+    }
+
+    user_data->request = request;
 
     struct aws_http_make_request_options request_options = {
         .self_size = sizeof(request_options),
         .on_response_headers = s_on_incoming_headers_fn,
         .on_response_header_block_done = NULL,
         .on_response_body = s_on_incoming_body_fn,
         .on_complete = s_on_stream_complete_fn,
-        .user_data = sts_web_identity_user_data,
+        .user_data = user_data,
         .request = request,
     };
 
-    stream = impl->function_table->aws_http_connection_make_request(
-        sts_web_identity_user_data->connection, &request_options);
+    stream = impl->function_table->aws_http_connection_make_request(user_data->connection, &request_options);
```

Comment:
We still have the problem where any error from here til the end leaves us with user_data holding a destroyed request.

Reply:
You are damn right. There are similar problems in other providers which trigger http request. I have changed them as well.